File: 74.search-a-2-d-matrix.java
/*
 * @lc app=leetcode id=74 lang=java
 *
 * [74] Search a 2D Matrix
 */

// @lc code=start
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rowNum = matrix.length, colNum = matrix[0].length;
        int left = 0, right = rowNum * colNum - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int midVal = matrix[mid / colNum][mid % colNum];
            if (midVal == target) {
                return true;
            } else if (target > midVal) {
                left = mid + 1;
            } else if (target < midVal) {
                right = mid - 1;
            }
        }

        return false;

    }
}
// @lc code=end

File: 74.search-a-2-d-matrix.py
#
# @lc app=leetcode id=74 lang=python3
#
# [74] Search a 2D Matrix
#

# @lc code=start

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        l, r = 0, m*n-1

        while l <= r:
            mid = (l+r)//2
            row, col = mid//n, mid % n
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] < target:
                l = mid+1
            else:
                r = mid-1

        return False


# @lc code=end
